File: src/modules/algorithm/seed/algorithms/hash-map/languages/python/separate-chaining-solution.py
class HashMapNode:
    def __init__(self, key, value):
        self.key = key
        self.value = value
        self.next = None

class HashMap:
    def __init__(self, capacity: int = 16):
        self.capacity = capacity
        self.size = 0
        self.buckets = [None] * capacity

    def _hash(self, key):
        if isinstance(key, int):
            return key % self.capacity
        key_string = str(key)
        hash_value = 0
        for i in range(len(key_string)):
            hash_value = (hash_value * 31 + ord(key_string[i])) % self.capacity
        return hash_value

    def put(self, key, value):
        index = self._hash(key)
        node = self.buckets[index]

        while node:
            if node.key == key:
                node.value = value
                return
            if node.next is None:
                break
            node = node.next

        new_node = HashMapNode(key, value)
        if self.buckets[index] is None:
            self.buckets[index] = new_node
        else:
            new_node.next = self.buckets[index]
            self.buckets[index] = new_node
        self.size += 1

    def get(self, key):
        index = self._hash(key)
        node = self.buckets[index]

        while node:
            if node.key == key:
                return node.value
            node = node.next

        return None

    def remove(self, key):
        index = self._hash(key)
        node = self.buckets[index]
        prev = None

        while node:
            if node.key == key:
                if prev is None:
                    self.buckets[index] = node.next
                else:
                    prev.next = node.next
                self.size -= 1
                return
            prev = node
            node = node.next

    def size(self):
        return self.size

    def is_empty(self):
        return self.size == 0 

File: src/modules/algorithm/seed/algorithms/hash-map/languages/go/separate-chaining-solution.go
package hashmap

// HashMapNode represents a node in the linked list for separate chaining.
type HashMapNode struct {
	Key   interface{}
	Value interface{}
	Next  *HashMapNode
}

// HashMap represents a hash map data structure.
type HashMap struct {
	capacity int
	size     int
	buckets  []*HashMapNode
}

// NewHashMap creates a new HashMap with the given capacity.
func NewHashMap(capacity int) *HashMap {
	return &HashMap{
		capacity: capacity,
		buckets:  make([]*HashMapNode, capacity),
	}
}

// hash calculates the hash code for the given key.
func (hm *HashMap) hash(key interface{}) int {
	hash := 0
	keyString, ok := key.(string)
	if ok {
		for _, char := range keyString {
			hash = (hash*31 + int(char)) % hm.capacity
		}
	} else {
		// Simple fallback for non-string keys (not ideal for production)
		hash = 0
	}
	return hash
}

// Put inserts a key-value pair into the hash map.
func (hm *HashMap) Put(key interface{}, value interface{}) {
	index := hm.hash(key)
	node := hm.buckets[index]

	for node != nil {
		if node.Key == key {
			node.Value = value
			return
		}
		node = node.Next
	}

	newNode := &HashMapNode{Key: key, Value: value, Next: hm.buckets[index]}
	hm.buckets[index] = newNode
	hm.size++
}

// Get retrieves the value associated with the given key.
func (hm *HashMap) Get(key interface{}) interface{} {
	index := hm.hash(key)
	node := hm.buckets[index]

	for node != nil {
		if node.Key == key {
			return node.Value
		}
		node = node.Next
	}

	return nil
}

// Remove deletes the key-value pair associated with the given key.
func (hm *HashMap) Remove(key interface{}) {
	index := hm.hash(key)
	node := hm.buckets[index]
	var prev *HashMapNode

	for node != nil {
		if node.Key == key {
			if prev == nil {
				hm.buckets[index] = node.Next
			} else {
				prev.Next = node.Next
			}
			hm.size--
			return
		}
		prev = node
		node = node.Next
	}
}

// Size returns the number of key-value pairs in the hash map.
func (hm *HashMap) Size() int {
	return hm.size
}

// IsEmpty returns true if the hash map is empty, false otherwise.
func (hm *HashMap) IsEmpty() bool {
	return hm.size == 0
} 